File: compiler/queries/highlight.go
package queries

import (
	"wipple/database"
	"wipple/nodes/types"
	"wipple/typecheck"
	"wipple/visit"
)

func HighlightType(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(struct{})) {
	if _, ok := database.GetFact[types.IsTypeFact](node); ok {
		if _, ok := node.(*types.TypeParameterNode); !ok {
			f(struct{}{})
		}
	}
}

func HighlightTrait(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(struct{})) {
	if resolved, ok := database.GetFact[visit.ResolvedFact](node); ok {
		for _, definition := range resolved.Definitions {
			if _, ok := definition.(*visit.TraitDefinition); ok {
				f(struct{}{})
				return
			}
		}
	}

	if defined, ok := database.GetFact[visit.DefinedFact](node); ok {
		if _, ok := defined.Definition.(*visit.TraitDefinition); ok {
			f(struct{}{})
		}
	}
}

func HighlightFunction(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(struct{})) {
	highlightTag(db, node, "function", f)
}

func HighlightTypeParameter(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(struct{})) {
	highlightTag(db, node, "parameter", f)
}

func highlightTag(db *database.Db, node database.Node, tag any, f func(struct{})) {
	if _, ok := database.GetFact[visit.ResolvedFact](node); !ok {
		return
	}

	if _, ok := database.GetFact[visit.DefinedFact](node); ok {
		return
	}

	typed, ok := database.GetFact[typecheck.TypedFact](node)
	if !ok || typed.Group == nil || len(typed.Group.Types) == 0 {
		return
	}

	if typed.Group.Types[0].Tag == tag {
		f(struct{}{})
	}
}

File: compiler/src/queries/highlight.rs
use crate::{
    nodes::{IsType, TypeParameterNode},
    queries::QueryCtx,
    typecheck::{ConstructedTypeTag, Typed},
    visit::{Defined, Definition, Resolved},
};

pub fn highlight_type(ctx: &QueryCtx<'_>, f: &mut dyn FnMut(())) {
    if ctx.db.contains::<IsType>(&ctx.node)
        && ctx.node.downcast_ref::<TypeParameterNode>().is_none()
    {
        f(());
    }
}

pub fn highlight_trait(ctx: &QueryCtx<'_>, f: &mut dyn FnMut(())) {
    let definition_node = ctx
        .db
        .get::<Resolved>(&ctx.node)
        .and_then(|Resolved { definitions, .. }| definitions.into_iter().next())
        .unwrap_or_else(|| ctx.node.clone());

    if let Some(Defined(Definition::Trait(_))) = ctx.db.get(&definition_node) {
        f(());
    }
}

pub fn highlight_function(ctx: &QueryCtx<'_>, f: &mut dyn FnMut(())) {
    highlight_tag(ctx, |tag| matches!(tag, ConstructedTypeTag::Function), f);
}

pub fn highlight_type_parameter(ctx: &QueryCtx<'_>, f: &mut dyn FnMut(())) {
    highlight_tag(
        ctx,
        |tag| matches!(tag, ConstructedTypeTag::Parameter(_)),
        f,
    );
}

fn highlight_tag(
    ctx: &QueryCtx<'_>,
    filter: impl Fn(&ConstructedTypeTag) -> bool,
    f: &mut dyn FnMut(()),
) {
    if ctx.db.get::<Resolved>(&ctx.node).is_none() {
        return;
    }

    if ctx.db.get::<Defined>(&ctx.node).is_some() {
        return;
    }

    let Some(Typed { group: Some(group) }) = ctx.db.get(&ctx.node) else {
        return;
    };

    if group.types.first().is_some_and(|ty| filter(&ty.tag)) {
        f(());
    }
}
